File: android/app/src/test/kotlin/com/kenneth/stockcalc/data/supabase/TradeMapperTest.kt
package com.kenneth.stockcalc.data.supabase

import com.kenneth.stockcalc.domain.model.StopLossEntry
import com.kenneth.stockcalc.domain.model.Trade
import com.kenneth.stockcalc.domain.model.TradeStatus
import kotlinx.datetime.Instant
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class TradeMapperTest {
    private val now = Instant.parse("2026-04-23T02:00:00Z")

    private val domain = Trade(
        id = "00000000-0000-0000-0000-000000000001",
        symbol = "AAPL",
        entryPrice = 100.0,
        shares = 10,
        initialStopLoss = 90.0,
        currentStopLoss = 92.0,
        targetPrice = 120.0,
        status = TradeStatus.OPEN,
        riskAmount = 100.0,
        createdAt = now,
        stopLossHistory = listOf(StopLossEntry(90.0, now, "初始止損")),
    )

    @Test
    fun `toDto maps camelCase to snake_case with non-numeric id dropped`() {
        val dto = TradeMapper.toDto(domain, userId = "user-1")
        assertEquals(null, dto.id)               // UUID string not numeric → id dropped for insert
        assertEquals("user-1", dto.userId)
        assertEquals("AAPL", dto.symbol)
        assertEquals(100.0, dto.entryPrice)
        assertEquals(10, dto.shares)
        assertEquals(90.0, dto.initialStopLoss)
        assertEquals(92.0, dto.currentStopLoss)
        assertEquals(120.0, dto.targetPrice)
        assertEquals("open", dto.status)
        assertEquals(1, dto.stopLossHistory.size)
    }

    @Test
    fun `toDto preserves numeric id for remote rows`() {
        val dto = TradeMapper.toDto(domain.copy(id = "26"), userId = "user-1")
        assertEquals(26L, dto.id)
    }

    @Test
    fun `toDomain maps numeric id to string and snake_case fields to camelCase`() {
        val dto = TradeDto(
            id = 26L,
            userId = "user-1",
            symbol = "0700.HK",
            entryPrice = 320.0,
            shares = 5,
            initialStopLoss = 310.0,
            currentStopLoss = 315.0,
            targetPrice = 340.0,
            status = "closed",
            riskAmount = 50.0,
            createdAt = "2026-04-23T02:00:00Z",
            stopLossHistory = emptyList(),
            exitPrice = 335.0,
            pnl = 75.0,
            closedAt = "2026-04-23T03:00:00Z",
        )
        val trade = TradeMapper.toDomain(dto)
        assertEquals("26", trade.id)
        assertEquals(TradeStatus.CLOSED, trade.status)
        assertEquals(335.0, trade.exitPrice)
        assertEquals("0700.HK", trade.symbol)
    }

    @Test
    fun `local dto round trip preserves uuid id`() {
        val local = TradeMapper.toLocalDto(domain)
        assertEquals("00000000-0000-0000-0000-000000000001", local.id)
        val back = TradeMapper.fromLocalDto(local)
        assertEquals(domain, back)
    }
}

File: src/lib/trades/tradeMapper.test.ts
import { describe, it, expect } from 'vitest';
import { dtoToTrade, tradeToDbRow } from './tradeMapper';
import type { TradeDto, Trade } from '@/types/trade';

const dto: TradeDto = {
  id: 'abc', user_id: 'u1', symbol: 'AAPL', entry_price: 150, shares: 128,
  initial_stop_loss: 145, current_stop_loss: 145, target_price: 165, status: 'open',
  risk_amount: 640, stop_loss_history: [{ price: 145, date: '2026-06-10T00:00:00Z', note: '初始止損' }],
  exit_price: null, pnl: null, closed_at: null, created_at: '2026-06-10T00:00:00Z',
};

describe('dtoToTrade', () => {
  it('maps snake_case columns to camelCase domain fields', () => {
    const t = dtoToTrade(dto);
    expect(t).toMatchObject({
      id: 'abc', symbol: 'AAPL', entryPrice: 150, shares: 128, initialStopLoss: 145,
      currentStopLoss: 145, targetPrice: 165, status: 'open', riskAmount: 640,
      createdAt: '2026-06-10T00:00:00Z', exitPrice: null, pnl: null, closedAt: null,
    });
    expect(t.stopLossHistory).toEqual(dto.stop_loss_history);
  });
  it('defaults null stop_loss_history to []', () => {
    expect(dtoToTrade({ ...dto, stop_loss_history: null }).stopLossHistory).toEqual([]);
  });
});

describe('tradeToDbRow', () => {
  it('maps domain fields to DB columns with user_id and null-coalesced money', () => {
    const trade: Trade = dtoToTrade(dto);
    const row = tradeToDbRow(trade, 'u1');
    expect(row).toMatchObject({
      user_id: 'u1', symbol: 'AAPL', entry_price: 150, shares: 128, initial_stop_loss: 145,
      current_stop_loss: 145, target_price: 165, status: 'open', risk_amount: 640,
      exit_price: null, pnl: null, closed_at: null,
    });
    expect(row.stop_loss_history).toEqual(trade.stopLossHistory);
    expect('id' in row).toBe(false);
    expect('created_at' in row).toBe(false);
  });
});
